generic-ec/src/multiscalar.rs

```diff
@@ -0,0 +1,406 @@
+/// Q_{k-2} &= 16 Q_{k-1} +& T_{1,s_{1,k-2}} &+ \dots &+ T_{n,s_{n,k-2}}& \\\\
+/// \vdots  &              &                 &        &                 & \\\\
+/// Q_1     &= 16 Q_2     +& T_{1,s_{1,1}}   &+ \dots &+ T_{n,s_{n,1}}  & \\\\
+/// Q = Q_0 &= 16 Q_1     +& T_{1,s_{1,0}}   &+ \dots &+ T_{n,s_{n,0}}  &
+/// \end{aligned}
+/// $$
+#[cfg(feature = "alloc")]
+pub struct Straus;
+
+#[cfg(feature = "alloc")]
+impl<E: Curve> MultiscalarMul<E> for Straus {
+    fn multiscalar_mul<S, P>(scalar_points: impl IntoIterator<Item = (S, P)>) -> Point<E>
+    where
+        S: AsRef<Scalar<E>>,
+        P: AsRef<Point<E>>,
+    {
+        let (mut scalars, points): (Vec<Radix16Iter<E>>, Vec<Point<E>>) = scalar_points
+            .into_iter()
+            .map(|(scalar, point)| (scalar.as_ref().as_radix16_be(), *point.as_ref()))
+            .unzip();
+        Self::multiscalar_mul_inner(&mut scalars, &points)
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl Straus {
+    fn multiscalar_mul_inner<E: Curve>(
+        scalars: &mut [Radix16Iter<E>],
+        points: &[Point<E>],
+    ) -> Point<E> {
+        if scalars.is_empty() {
+            return Point::zero();
+        }
+
+        // table[i] = [point_i, 2 * point_i, ..., 15 * point_i]
+        let table = points
+            .iter()
+            .map(|point_i| {
+                iter::successors(Some(*point_i), |point| Some(point + point_i))
+                    .take(15)
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+
+        // `serialized_len` is amount of radix256-digits. We multiply it by 2 and get
+        // amount of radix16-digits
+        let num_digits = 2 * Scalar::<E>::serialized_len();
+
+        let mut sum = Point::<E>::zero();
+        for j in 0..num_digits {
+            // partial_sum = \sum_i s_{i,k} P_i where `k` is index of the most significant
+            // unprocessed coefficient
+            let partial_sum = scalars
+                .iter_mut()
+                .map(|radix16| {
+                    #[allow(clippy::expect_used)]
+                    radix16
+                        .next()
+                        .expect("there must be next radix16 available")
+                })
+                .enumerate()
+                .flat_map(|(i, v)| {
+                    // We need to calculate `v * P_i`
+                    if v == 0 {
+                        // P_i * 0 = 0, we don't include it into the sum
+                        None
+                    } else {
+                        debug_assert!(v < 16);
+                        Some(table[i][usize::from(v) - 1])
+                    }
+                })
+                .sum::<Point<E>>();
+
+            if j == 0 {
+                sum = partial_sum
+            } else {
+                // sum = 16 * sum + partial_sum
+                let sum_at_16 = sum.double().double().double().double();
+                sum = sum_at_16 + partial_sum;
+            }
```

Comment:
Medium nitpick: I think it would be slightly easier to understand if `sum = if j == 0 { sum } else { sum.double().double().double().double() };` was at the top of the loop, similar to how it's usually done with polynomial values for example.

Reply:
I think it would be more confusing. It should be obvious that we shift `sum` by 3 bits so we could write our result. If we don't do that when we write our result, this is confusing